### PR TITLE
Add autoExpand option to expand the bar to the window width

### PR DESCRIPTION
**Pull request by rottmann**
With the option `autoExpand: true` the bar will be expanded to the window width.

Change of custom context parameters (change of length) will be considered.

### CODE REVIEW

File: lib/format-bar.js [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
-            options.barIncompleteString.substr(0, incompleteSize);
-}
\ No newline at end of file
+    let columns = options.barsize;
+    if (options.autoExpand) {
+      columns = (options.stream && options.stream.columns) ? options.stream.columns : columns;
+      columns = columns - options.restWidth - platformSpace;
+
+      // dynamic width handling, on window resize and change of other context parameters
+      options.barCompleteString = (new Array(columns + 1 ).join(options.barCompleteChar));
```
Comment (AndiDittrich) [inline lib/format-bar.js:15]: The strings should be pre-computed for performance reasons

File: lib/formatter.js [hunk 2]
```diff
@@ -53,6 +51,10 @@ module.exports = function defaultFormatter(options, params, payload){
         return match;
     });
 
+    // replace bar
+    const bar = formatBar(params.progress, Object.assign({}, options, { restWidth: s.length - 5 })); // 5 = {bar}
```
Comment (AndiDittrich) [inline lib/formatter.js:55]: why is it moved outside the standard placeholder processing ?

File: lib/options.js [hunk 3]
```diff
@@ -90,6 +90,13 @@ module.exports = {
         // autopadding character - empty in case autopadding is disabled
         options.autopaddingChar = options.autopadding ? mergeOption(opt.autopaddingChar, '   ') : '';
 
+        // auto expand bar width to visible window width
+        options.autoExpand = mergeOption(opt.autoExpand, false);
+
+        // necessary for auto expand and dyanmic window resize
+        options.barCompleteChar = opt.barCompleteChar || '=';
```
Comment (AndiDittrich) [inline lib/options.js:97]: not required in case the pre-computed string sequences are used

File: lib/format-bar.js [hunk 4]
```diff
@@ -1,11 +1,26 @@
+let platformSpace = 0;
+if (process.platform === 'win32'){
```
Comment (AndiDittrich) [inline lib/format-bar.js:2]: please add some information why this is necessary. what's about linux/osx/bsd ?

File: README.md [hunk 5]
```diff
@@ -231,6 +231,7 @@ The following options can be changed
 - `stopOnComplete` (type:boolean) - automatically call `stop()` when the value reaches the total (default: false)
 - `clearOnComplete` (type:boolean) - clear the progress bar on complete / `stop()` call (default: false)
 - `barsize` (type:int) - the length of the progress bar in chars (default: 40)
+- `autoExpand` (type:boolean) - ignores `barsize` and expands the bar to the window width (default: false)
```
Comment (AndiDittrich) [inline README.md:234]: maybe `barsize`='auto' is more convenient ?

**Review by AndiDittrich — CHANGES_REQUESTED** [review]
Hi @rottmann ,

thanks for your contribution. i've added a few comments to your code regarding some possible issues which has to be resolved before merging

File: lib/format-bar.js (hunk 1 above)
Comment (rottmann) [reply lib/format-bar.js:15]: Could only pre compute it with a huge length, cause we don't know big the screen is. I think performance is not really a problem of the progressbar.

File: lib/formatter.js (hunk 2 above)
Comment (rottmann) [reply lib/formatter.js:55]: We need to know the length of the remaining string befor the bar could be computed.

File: lib/options.js (hunk 3 above)
Comment (rottmann) [reply lib/options.js:97]: See above, on static view you are right, but we don't knwo the maximum window size, the window could be resized.

File: lib/format-bar.js (hunk 4 above)
Comment (rottmann) [reply lib/format-bar.js:2]: Saw this on progress https://github.com/visionmedia/node-progress/blob/0790207ef077cbfb7ebde24a1dd9895ebf4643e1/lib/node-progress.js#L152, i did not test it, but i think they use it for a reason.

File: README.md (hunk 5 above)
Comment (rottmann) [reply README.md:234]: Yes possible too, or set barsize to 0 or Infinity

**Comment by AndiDittrich** [comment]
currently, i'm not sure if it's a good idea to merge it in it's current state. especially the platform compatibility should be tested - a lot of issues in the past are raised due to too large columns or strange behaviour of specific shells.

File: lib/format-bar.js (hunk 4 above)
Comment (rottmann) [reply lib/format-bar.js:2]: I run test on Windows 10 cmd line, the code is not necessary. I can not reproduce why node-progress uses the 1 space correction.

Comment (rottmann) [reply lib/format-bar.js]: Correction, under win7 cmd line it is not necessary, win 10 cmd line need the space!

Comment (AndiDittrich) [reply lib/format-bar.js:2]: mostly `cmd.exe` and `powershell.exe` shows a different behaviour regarding ansii escape sequences...this also has to be tested within all major version. therefore we should look for another solution...

**Comment by rottmann** [comment]
If not merged, no problem. But could you then export the default formatter and functions (formatter.js, format-value.js, format-time.js, ...) in cli-progress-js, so i can write a custom formatter that uses the default functions.

**Pull request closed without merging**